src/client/Microsoft.Identity.Client/Platforms/netdesktop/NetDesktopCryptographyManager.cs

```diff
@@ -79,16 +84,25 @@ public byte[] SignWithCertificate(string message, X509Certificate2 certificate)
             byte[] messageBytes = Encoding.UTF8.GetBytes(message);
 
 #if NET45
-            var rsa = GetCryptoProviderForSha256_Net45(certificate);
+            var rsaCryptoProvider = GetCryptoProviderForSha256_Net45(certificate);
             using (var sha = new SHA256Cng())
             {
-                return rsa.SignData(messageBytes, sha);
+                var signedData = rsaCryptoProvider.SignData(messageBytes, sha);
+                // Cache only valid RSA crypto providers
+                s_certificateToRsaCspMap[certificate.Thumbprint] = rsaCryptoProvider;
+                return signedData;
             }
 #else
-            using (var key = certificate.GetRSAPrivateKey())
+            if (!s_certificateToRsaMap.TryGetValue(certificate.Thumbprint, out RSA rsa))
             {
-                return key.SignData(messageBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+                if (s_certificateToRsaMap.Count >= s_maximumMapSize)
+                    s_certificateToRsaMap.Clear();
+
+                s_certificateToRsaMap[certificate.Thumbprint] = certificate.GetRSAPrivateKey();
```

Comment:
@GeoK @brentschmaltz
Do we need to check here if RSA can sign successfully before caching?